<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Example: Modeling regional variation in mutation and recombination &#8212; fwdpy 0.0.4-rc3 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0.4-rc3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Temporal sampling" href="../examples/temporalSampling.html" />
    <link rel="prev" title="Complex demographic operations" href="demography.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="example-modeling-regional-variation-in-mutation-and-recombination">
<h1>Example: Modeling regional variation in mutation and recombination<a class="headerlink" href="#example-modeling-regional-variation-in-mutation-and-recombination" title="Permalink to this headline">¶</a></h1>
<p>Several of the simulation routines allow the details of the mutation and recombination models to vary along a &#8220;sequence&#8221; or &#8220;region&#8221;.  A user is able to specify the details of such variation by passing <em>lists</em> to package functions.  For example, you are able to:</p>
<ul class="simple">
<li>Vary the neutral mutation rate along a sequence.</li>
<li>Vary the distribution of selection coefficients (and the dominance associated with selected mutations) along a sequence.</li>
<li>Vary the recombination rate along a sequence.</li>
</ul>
<p>The implementation of such variation along a region is <em>discrete</em>.  A region is specified by a beginning, and end, and a weight, plus any additional data required to specify selection coefficients, dominance, etc.</p>
<div class="section" id="background">
<h2>Background<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h2>
<p>The models are parameterized through Python&#8217;s &#8220;new-style&#8221; class system.</p>
<div class="section" id="mutation-rates-recombination-rates-and-a-weighting-system">
<h3>Mutation rates, recombination rates, and a weighting system<a class="headerlink" href="#mutation-rates-recombination-rates-and-a-weighting-system" title="Permalink to this headline">¶</a></h3>
<p>A simulation will typically have a mutation rate, <img class="math" src="../_images/math/6021866aa7cf262fb9df5943c3dc0025fab9acee.png" alt="\\mu"/>, which represents the mean of a Poisson number of mutations per gamete per generation), and a recombination rate, <img class="math" src="../_images/math/eaa6ad49a7f78fe5a13b486690163bf2dc7e3e60.png" alt="r"/>, which again is the mean of Poisson number of crossover events (per diploid, per generation).  These parameters are the _total_ rates across an entire simulated region.  Variation in these parameters along the region are affected by a set of positions coupled with &#8220;weights&#8221;, which the user specifies using S4 classes.</p>
<p>The base class: <a class="reference internal" href="../index.html#fwdpy.fwdpy.Region" title="fwdpy.fwdpy.Region"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Region</span></code></a></p>
<p>A <a class="reference internal" href="../index.html#fwdpy.fwdpy.Region" title="fwdpy.fwdpy.Region"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Region</span></code></a> is an Python class with the following members:</p>
<ul class="simple">
<li><img class="math" src="../_images/math/57c9d14bb082716df9000146882ce365335d08f1.png" alt="b"/>, which is the beginning/start of the region. The type is &#8220;float&#8221;.</li>
<li><img class="math" src="../_images/math/78a06cf7a34f39fde71fbba0d5539baae7c72d4e.png" alt="e"/>, which is the end/stop of the region. The type is &#8220;float&#8221;.</li>
<li><img class="math" src="../_images/math/ecd1ee2a1cd226b40c37e079aca62398d4b774f5.png" alt="w"/>, which is a weighting factor associated with the region. The type is &#8220;float&#8221;.</li>
</ul>
<p>The members are used to inform the C++ code about the relative abundance of new mutations or recombination events will occur in what region.  Briefly, the number of events that occur in region <img class="math" src="../_images/math/df0deb143e5ac127f00bd248ee8001ecae572adc.png" alt="i"/> are proportional to <img class="math" src="../_images/math/94f17712f05697456c643580aa94b5526cce572e.png" alt="w_i/\sum_i w"/>, <em>i.e</em>, the weight assigned to region <img class="math" src="../_images/math/df0deb143e5ac127f00bd248ee8001ecae572adc.png" alt="i"/> divided by the sum of weights assigned to all regions.  The weights for mutation events and for recombination events are considered separately.  Thus, in order to model a correlation between mutational processes and recombination, it is up to the user to generate regions whose weights are correlated.</p>
<p>fwdpy allows the <img class="math" src="../_images/math/ecd1ee2a1cd226b40c37e079aca62398d4b774f5.png" alt="w"/> slot to be interpreted in one of two ways:</p>
<ul class="simple">
<li>It is <em>not</em>  affected by the length of region.  Interally, the weight assigned is simply <img class="math" src="../_images/math/ecd1ee2a1cd226b40c37e079aca62398d4b774f5.png" alt="w"/>.</li>
<li>It is affected by the length of a region <img class="math" src="../_images/math/af3730c6fa136636e2e0a22728f5fac1ff96ebf7.png" alt="(e - b)"/>.</li>
</ul>
<p>These two options are determined by arguments to class constructors, which we will see in examples below.  The latter is the default.</p>
<p>These two approaches allow for considerable modeling flexibility.  For example, the latter approach allows <img class="math" src="../_images/math/ecd1ee2a1cd226b40c37e079aca62398d4b774f5.png" alt="w"/> to be interpreted as a &#8220;per base-pair&#8221; rate.  Imagine that you wanted to simulate variation in recombination along discrete 100 kilobase chunks, and the rate of crossing-over <em>per base pair</em> increases in each chunk, and includes an initial chunk with no recombination:</p>
<ol class="arabic simple">
<li>start=1,stop= <img class="math" src="../_images/math/673ace6d151bdca7a2437fa0e74c58e7ec22194f.png" alt="10^5"/>, <img class="math" src="../_images/math/73d6cb47a91348bfa91e2497a34e96b253f3fdc1.png" alt="r_{bp}=0"/></li>
<li>start= <img class="math" src="../_images/math/673ace6d151bdca7a2437fa0e74c58e7ec22194f.png" alt="10^5"/>,stop= <img class="math" src="../_images/math/2f726ccc9d0d1833a5fad6fff788d5e8acc09963.png" alt="2 \times 10^5"/>, <img class="math" src="../_images/math/b04be388a17a1a54c35baa0a57e0aa1ac052c1a8.png" alt="r_{bp}=10^{-8}"/></li>
<li>start= <img class="math" src="../_images/math/2f726ccc9d0d1833a5fad6fff788d5e8acc09963.png" alt="2 \times 10^5"/>,stop= <img class="math" src="../_images/math/2156fa7c75e08c0738d373b792485fabc312e133.png" alt="3 \times 10^5"/>, <img class="math" src="../_images/math/bc8e48df3805bfddcc6b78d9ac6621958ceb5b23.png" alt="r_{bp}=10^{-7}"/></li>
</ol>
<p>This model boils down to the relative number of crossing overs per region occuring in the ratio <img class="math" src="../_images/math/f61c4f78c19a25fb93e39af05f49108631c9de13.png" alt="0 : 10^{-8} : 10^{-7}"/>.  This is easily represented using fwdpy&#8217;s classes:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">print_function</span>
<span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="n">recRegions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="n">e5</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">1</span><span class="n">e5</span><span class="p">,</span><span class="mi">2</span><span class="n">e5</span><span class="p">,</span><span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">8</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="n">e5</span><span class="p">,</span><span class="mi">3</span><span class="n">e5</span><span class="p">,</span><span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">7</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">recRegions</span><span class="p">:</span>
    <span class="nb">print</span> <span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">beg</span> <span class="o">=</span> <span class="mf">1.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">100000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.000000000</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">100000.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">200000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.001000000</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">200000.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">300000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.010000000</span>
</pre></div>
</div>
<p>For this hypothetical example, the region lengths are all identical, and
thus an equivalent specification would be this:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">recRegions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="n">e5</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="kc">False</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">1</span><span class="n">e5</span><span class="p">,</span><span class="mi">2</span><span class="n">e5</span><span class="p">,</span><span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">8</span><span class="p">,</span><span class="kc">False</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="n">e5</span><span class="p">,</span><span class="mi">3</span><span class="n">e5</span><span class="p">,</span><span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">7</span><span class="p">,</span><span class="kc">False</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">recRegions</span><span class="p">:</span>
    <span class="nb">print</span> <span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">beg</span> <span class="o">=</span> <span class="mf">1.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">100000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.000000000</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">100000.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">200000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.000000010</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">200000.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">300000.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.000000100</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="specific-examples">
<h2>Specific examples<a class="headerlink" href="#specific-examples" title="Permalink to this headline">¶</a></h2>
<div class="section" id="mutations-not-affecting-fitness-neutral-mutations">
<h3>Mutations not affecting fitness (&#8220;neutral&#8221; mutations)<a class="headerlink" href="#mutations-not-affecting-fitness-neutral-mutations" title="Permalink to this headline">¶</a></h3>
<p>You specify regions where neutral mutations arise via the class <a class="reference internal" href="../index.html#fwdpy.fwdpy.Region" title="fwdpy.fwdpy.Region"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Region</span></code></a>.  A region has a beginning, end, and a weight Thus, the following list would specify that 100% of neutral mutations occur on the continuous interval [0,1):</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="n">neutralRegions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
</pre></div>
</div>
<p>The beginning and end positions can be whatever you like:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1">#With a weight of 1, we&#39;re just rescaling the position here.</span>
<span class="n">neutralRegions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
</pre></div>
</div>
<p>To specify variation in the netural mutation process along a sequence,
combine multiple regions in your list:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1">#If coupled=False for the second region, the effect would be that region2&#39;s mutation rate per base pair is 10x less than region 1!!</span>
<span class="n">neutralRegions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">end</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">end</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">coupled</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>
</pre></div>
</div>
<p>Internally, the total &#8220;mutational weight&#8221; of the first region will be a
function of its length, which is 1(1-0)=1. The second region&#8217;s total
weight will be 1*(12-2)=10, and it will have 10xas many new mutations
arising as the first region.</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1">#Let&#39;s see what happens if we set coupled=False:</span>
<span class="n">neutralRegions2</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="n">end</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="n">beg</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">end</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">weight</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">)]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The set with coupled=True:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">neutralRegions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The set with coupled=False:&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">neutralRegions2</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">The</span> <span class="nb">set</span> <span class="k">with</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">True</span><span class="p">:</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">0.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">1.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">1.000000000</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">2.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">12.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">10.000000000</span>
<span class="n">The</span> <span class="nb">set</span> <span class="k">with</span> <span class="n">coupled</span><span class="o">=</span><span class="kc">False</span><span class="p">:</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">0.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">1.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">1.000000000</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">2.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">12.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">1.000000000</span>
</pre></div>
</div>
<p>See the difference in the above? (Look at the &#8220;weight&#8221; term in the
second line of each set.)</p>
</div>
<div class="section" id="mutations-affecting-fitness">
<h3>Mutations affecting fitness<a class="headerlink" href="#mutations-affecting-fitness" title="Permalink to this headline">¶</a></h3>
<p>Type types of mutations affecting fitness that we consider will have two parameters associated with them:</p>
<ul class="simple">
<li><img class="math" src="../_images/math/63751cb2e98ba393b0f22e45ca127c3cebb61487.png" alt="s"/>, the selection coefficient</li>
<li><img class="math" src="../_images/math/293fb39e1b93282c804a86186e721b32f829f1b2.png" alt="h"/>, the effect of the mutation in a heterozygote (a.k.a. the &#8220;dominance&#8221; of the mutation).</li>
</ul>
<p>In a simulation, we may place a distribution on either <img class="math" src="../_images/math/63751cb2e98ba393b0f22e45ca127c3cebb61487.png" alt="s"/> itself or on the scaled selection parameter <img class="math" src="../_images/math/d821c81b7842fa935ffd6009086beed7759120a9.png" alt="\alpha = 2Ns"/>.  These two methods are represented by the class <a class="reference internal" href="../index.html#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a>.  These classes contain/extend the <code class="xref py py-class docutils literal"><span class="pre">Region</span></code> class described above, and thus inherit their members.  <code class="xref py py-class docutils literal"><span class="pre">Sregion</span></code> adds <img class="math" src="../_images/math/293fb39e1b93282c804a86186e721b32f829f1b2.png" alt="h"/>, which is the dominance of a mutation, and then classes extending <code class="xref py py-class docutils literal"><span class="pre">Sregion</span></code> add details about the distribution of fitness effects.  These classes are:</p>
<ul class="simple">
<li><a class="reference internal" href="../index.html#fwdpy.fwdpy.ConstantS" title="fwdpy.fwdpy.ConstantS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.ConstantS</span></code></a></li>
<li><a class="reference internal" href="../index.html#fwdpy.fwdpy.UniformS" title="fwdpy.fwdpy.UniformS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.UniformS</span></code></a></li>
<li><a class="reference internal" href="../index.html#fwdpy.fwdpy.GammaS" title="fwdpy.fwdpy.GammaS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.GammaS</span></code></a></li>
<li><a class="reference internal" href="../index.html#fwdpy.fwdpy.GaussianS" title="fwdpy.fwdpy.GaussianS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.GaussianS</span></code></a></li>
</ul>
</div>
<div class="section" id="crossover-rate-variation">
<h3>Crossover rate variation<a class="headerlink" href="#crossover-rate-variation" title="Permalink to this headline">¶</a></h3>
<p>Just like neutral mutations, intervals with different crossover rates are specified by different <code class="xref py py-class docutils literal"><span class="pre">Region</span></code> objects.  Let&#8217;s set up the following concrete example:</p>
<ul class="simple">
<li>A region where crossovers occur between positions [0,1)</li>
<li>Positions [0,0.45) and [0.55,1) have uniform recombintion rates</li>
<li>Positions [0.45,0.55) are a recombination hotspot with 100x the background intensity (per &#8220;base pair&#8221;).</li>
</ul>
<p>The above model can be represented as:</p>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1">#recrate[2] is the hotspot:</span>
<span class="n">recrates</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mf">0.</span><span class="p">,</span><span class="mf">0.45</span><span class="p">,</span><span class="mf">1.</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mf">0.55</span><span class="p">,</span><span class="mf">1.</span><span class="p">,</span><span class="mf">1.</span><span class="p">,),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mf">0.45</span><span class="p">,</span><span class="mf">0.55</span><span class="p">,</span><span class="mf">100.</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">recrates</span><span class="p">:</span>
    <span class="nb">print</span> <span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">beg</span> <span class="o">=</span> <span class="mf">0.000000000</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">0.449999988</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.449999988</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">0.550000012</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">1.000000000</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">0.449999988</span>
<span class="n">beg</span> <span class="o">=</span> <span class="mf">0.449999988</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="mf">0.550000012</span><span class="p">,</span> <span class="n">weight</span> <span class="o">=</span> <span class="mf">10.000002384</span>
</pre></div>
</div>
<p><strong>Please note:</strong> the apparent &#8216;slop&#8217; that you see above (<em>e.g.</em>, the output looks oddly rounded vis-a-vis the input) does not appear to be passed on to the C++ internals, which is where it really matters.  This appears to simply be a display issue.</p>
<p>Internally, this is what will happen to the above input:</p>
<ul class="simple">
<li>The total weight on the first region will be <img class="math" src="../_images/math/a44f7f39755c1df4126b67a4d7ebe0b7a3297ddf.png" alt="w = w \times (e-b) = 1\times(0.45-0) = 0.45"/></li>
<li>The weight on the second region will be <img class="math" src="../_images/math/69822538257813d9277137c4b05d03411e632684.png" alt="1\times(1-0.55) = 0.45"/></li>
<li>The weight on the hotspot will be <img class="math" src="../_images/math/615da1efec47c25a5a6293627dcbe1beafc5fc85.png" alt="100\times(0.55-0.45) = 10"/></li>
</ul>
<p>This gives us what we want: the hotspot is 100x hotter &#8220;per base&#8221;, and is 10% of the total region in length.  We therefore expect 10x as many crossovers in that region as in the flanking regions.</p>
</div>
</div>
<div class="section" id="how-to-set-up-a-model">
<h2>How to set up a model<a class="headerlink" href="#how-to-set-up-a-model" title="Permalink to this headline">¶</a></h2>
<p>When setting up a model, it is important that you think in terms of conditional probabilities.  In other words, if the total rate to neutral variants is <img class="math" src="../_images/math/f2e6fcf65a9eb34023e8a22076dfd9550eb9f3ee.png" alt="\mu_n"/>, then the weights passed along to a function have the interpretations &#8220;Given that a neutral mutation occurs, the probability that it occurs in a certain interval is <img class="math" src="../_images/math/a59f68a4202623bb859a7093f0316bf466e6f75d.png" alt="x"/>, where <img class="math" src="../_images/math/a59f68a4202623bb859a7093f0316bf466e6f75d.png" alt="x"/> is determined by the relative weight assigned to an interval.</p>
<p>The &#8216;weights&#8217; that you assign are <em>relative</em> and need not sum to 1.  Each weight must be <img class="math" src="../_images/math/f37e6efc0713a22751c035108f69071fd8e8cf17.png" alt="\geq 0"/>, though.</p>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="c1">##Some basic parameters</span>
<span class="n">N</span><span class="o">=</span><span class="mi">1000</span>
<span class="n">theta</span><span class="o">=</span><span class="mf">100.0</span>
<span class="n">rho</span><span class="o">=</span><span class="mf">100.0</span>
<span class="c1">##All neutral muts are [0,1)</span>
<span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span> <span class="p">]</span>
<span class="c1">#Selected mutations.  All are additive, to keep this example simple.</span>
<span class="c1">##Strongly deleterious mutations to the &quot;left&quot;</span>
<span class="c1">##Weaker mutations (2Ns = 10 on average) to the &quot;right&quot;</span>
<span class="c1">## 1% of selected mutations will be positively selected</span>
<span class="c1">## and uniform throughout the region.  The distribution</span>
<span class="c1">## of s will be exponential with mean 1e-3</span>
<span class="n">smodels</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.99</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.99</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span><span class="o">-</span><span class="mi">10</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="c1">##Recombination models--10x hotspot in the middl</span>
<span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mf">0.45</span><span class="p">,</span><span class="mf">0.55</span><span class="p">,</span><span class="mi">10</span><span class="p">)]</span>
<span class="c1">#set up list of population sizes,</span>
<span class="c1">#which are NumPy arrays of ints</span>
<span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">N</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">10</span><span class="o">*</span><span class="n">N</span><span class="p">)</span>
<span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="n">theta</span><span class="o">/</span><span class="p">(</span><span class="mi">4</span><span class="o">*</span><span class="n">N</span><span class="p">),</span><span class="mf">0.1</span><span class="o">*</span><span class="n">theta</span><span class="o">/</span><span class="p">(</span><span class="mi">4</span><span class="o">*</span><span class="n">N</span><span class="p">),</span><span class="n">rho</span><span class="o">/</span><span class="p">(</span><span class="mi">4</span><span class="o">*</span><span class="n">N</span><span class="p">),</span><span class="n">nregions</span><span class="p">,</span><span class="n">smodels</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="c1">#Take a sample of size n = 10 from the population via list comprehension</span>
<span class="n">popSample</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">get_samples</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
</pre></div>
</div>
<p>If we pass <em>the same neutral and selected mutation rates to evolve.regions</em>, then the above model satisfies:</p>
<ul class="simple">
<li>The total number of mutations occurring in Exon 2 is 2x the number occuring in Exon 1.</li>
<li>Within an exon, 3/4 of all new mutations are deleterious.</li>
</ul>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="c1">#Let&#39;s convert from tuples to pandas DataFrames.</span>
<span class="c1">#Ideally, one would further split each tuple element into a list,</span>
<span class="c1">#but this example let&#39;s us get the point...</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="n">neutralMuts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">popSample</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span><span class="s1">&#39;genotypes&#39;</span><span class="p">])</span>
<span class="n">selectedMuts</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span><span class="n">popSample</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span><span class="s1">&#39;genotypes&#39;</span><span class="p">])</span>
</pre></div>
</div>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">neutralMuts</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>        <span class="n">pos</span>                                          <span class="n">genotypes</span>
<span class="mi">0</span>  <span class="mf">0.006588</span>  <span class="mf">0000000000000000000000001000010000000000000000.</span><span class="o">..</span>
<span class="mi">1</span>  <span class="mf">0.012971</span>  <span class="mf">1000000000010101000001000010101000000000000010.</span><span class="o">..</span>
<span class="mi">2</span>  <span class="mf">0.014566</span>  <span class="mf">0110000010000010000010010000000111001000010000.</span><span class="o">..</span>
<span class="mi">3</span>  <span class="mf">0.014729</span>  <span class="mf">0000000000000000000000000000000000000000100000.</span><span class="o">..</span>
<span class="mi">4</span>  <span class="mf">0.020637</span>  <span class="mf">0000010000000000000000000000000000000000000000.</span><span class="o">..</span>
</pre></div>
</div>
<div class="code python highlight-default"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">selectedMuts</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
<div class="highlight-default"><div class="highlight"><pre><span></span>        <span class="n">pos</span>                                          <span class="n">genotypes</span>
<span class="mi">0</span>  <span class="mf">1.016680</span>  <span class="mf">0101110011001110000011010011011001010000011100.</span><span class="o">..</span>
<span class="mi">1</span>  <span class="mf">1.192438</span>  <span class="mf">1010001100110001111100101100100110101111100011.</span><span class="o">..</span>
<span class="mi">2</span>  <span class="mf">1.197182</span>  <span class="mf">0101110011001110000011010011011001010000011100.</span><span class="o">..</span>
<span class="mi">3</span>  <span class="mf">1.365021</span>  <span class="mf">0101110011001110000011010011011001010000011100.</span><span class="o">..</span>
<span class="mi">4</span>  <span class="mf">1.402332</span>  <span class="mf">1010001100110001111100101100100110101111100011.</span><span class="o">..</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Example: Modeling regional variation in mutation and recombination</a><ul>
<li><a class="reference internal" href="#background">Background</a><ul>
<li><a class="reference internal" href="#mutation-rates-recombination-rates-and-a-weighting-system">Mutation rates, recombination rates, and a weighting system</a></li>
</ul>
</li>
<li><a class="reference internal" href="#specific-examples">Specific examples</a><ul>
<li><a class="reference internal" href="#mutations-not-affecting-fitness-neutral-mutations">Mutations not affecting fitness (&#8220;neutral&#8221; mutations)</a></li>
<li><a class="reference internal" href="#mutations-affecting-fitness">Mutations affecting fitness</a></li>
<li><a class="reference internal" href="#crossover-rate-variation">Crossover rate variation</a></li>
</ul>
</li>
<li><a class="reference internal" href="#how-to-set-up-a-model">How to set up a model</a><ul>
<li><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="demography.html" title="previous chapter">Complex demographic operations</a></li>
      <li>Next: <a href="../examples/temporalSampling.html" title="next chapter">Temporal sampling</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pages/regions.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, KRT.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
      |
      <a href="../_sources/pages/regions.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>