<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Estimating cumulative in simulations of quantitative traits. &#8212; fwdpy 0.0.4-rc3 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0.4-rc3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Overview" href="plugins.html" />
    <link rel="prev" title="Copying populations in memory" href="../examples/copypops.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="estimating-cumulative-in-simulations-of-quantitative-traits">
<h1>Estimating cumulative <img class="math" src="../_images/math/aeb4b64115fe29aa42095dc692f5b5fb43dac071.png" alt="V(A)"/> in simulations of quantitative traits.<a class="headerlink" href="#estimating-cumulative-in-simulations-of-quantitative-traits" title="Permalink to this headline">¶</a></h1>
<p>When simulating quantitative traits, one is often interested in the relationship between mutation frequency and contribution to overall genetic variance.  Specifically, a plot where the x axis is mutation frequency and the y axis is the cumulative relative/percent contribution to genetic variance serves as a summary of the &#8220;genetic architecture&#8221; of a trait that has been used by several authors.</p>
<p>Under additive models, it is tempting to use the equation from the methods section of Simons <em>et al</em> (2014), which provides an explicit formula for generating such curves.  Similar formulas exist in Eyre-Walker (2010).  However, <em>they are incorrect when the assumption of independent sites is violated</em>, because the genetic background on which a site finds itself is a function of both linkage and the effect sizes of mutations, and hence not randomized each generation.  Thus, it is incorrect to apply these formulas to simulations from recombining regions.  In other words, if such curves were used to estimate the total additive variance (<img class="math" src="../_images/math/aeb4b64115fe29aa42095dc692f5b5fb43dac071.png" alt="V(A)"/>) under an additive model, from simulations where sites affecting the trait are partially-linked, then the result would be different from the actual total genetic variance (<img class="math" src="../_images/math/dc1280bd1774242a261602cf2b9e08dab050b7bb.png" alt="V(G)"/>).</p>
<p>In order to generate these curves from simulations with linkage, we turn to an approach based on explicit linear regression and estimation of effect sizes within frequency classes based on a Type I sum of squares.  We first applied this procedure in Sanjak <em>et al.</em>, and it is re-implemented here.</p>
<p>The procedure is as follows:</p>
<ul class="simple">
<li><img class="math" src="../_images/math/a59f68a4202623bb859a7093f0316bf466e6f75d.png" alt="x"/> is a vector of genetic values for each diploid in the population</li>
<li><img class="math" src="../_images/math/dee63237674afc3154c2e33b7b68d67f85f2ee0a.png" alt="G"/> is a matrix with values 0, 1, or 2, representing the number of copies of each mutation contributing to <img class="math" src="../_images/math/dc1280bd1774242a261602cf2b9e08dab050b7bb.png" alt="V(G)"/>.  For a population of <img class="math" src="../_images/math/f4170ed8938b79490d8923857962695514a8e4cb.png" alt="N"/> diploids with <img class="math" src="../_images/math/edba97b4c0d864d26e92ea7ea73767fa38eef3f7.png" alt="m"/> segregating variants affecting the trait, the matrix has <img class="math" src="../_images/math/f4170ed8938b79490d8923857962695514a8e4cb.png" alt="N"/> rows and <img class="math" src="../_images/math/edba97b4c0d864d26e92ea7ea73767fa38eef3f7.png" alt="m"/> columns.</li>
</ul>
<p>The matrix <img class="math" src="../_images/math/dee63237674afc3154c2e33b7b68d67f85f2ee0a.png" alt="G"/> is sorted as follows:</p>
<ul class="simple">
<li>From left to right, columns are sorted in <em>descending</em> order of column sums (mutation frequency)</li>
<li>Within each frequency group, columns are sorted in <em>descending</em> order based on the absolute value of effect sizes.</li>
</ul>
<p>The justification for this sorting order comes from fairly standard considerations, namely that the contribution to <img class="math" src="../_images/math/1eb0322b22c0832663929e2a009e9b2ad0236b2e.png" alt="V(A) \propto 2pqa^2"/>, where <img class="math" src="../_images/math/bd9d9c7512c898e0a254a0882184197bc1bd9c9f.png" alt="a"/> is the difference in trait value between genotypes <em>AA</em> and <em>aa</em>.  Note, however, that this is a modeling choice, and that it would also be justified to sort the matrix according to <img class="math" src="../_images/math/14edd86cec302aaf3002282ea33d6d9f301cb35e.png" alt="2pqa^2"/>.  Maybe we will add that as an option in the future.</p>
<p>If <img class="math" src="../_images/math/dee63237674afc3154c2e33b7b68d67f85f2ee0a.png" alt="G"/> contains redundant columns, they are collapsed into a single column.</p>
<p>We regress <img class="math" src="../_images/math/a59f68a4202623bb859a7093f0316bf466e6f75d.png" alt="x"/> onto <img class="math" src="../_images/math/dee63237674afc3154c2e33b7b68d67f85f2ee0a.png" alt="G"/> using QR decomposition, and obtain the sum of squares for each predictor as well as the residual sum of squares.  The total sum of squares (TSS) is the sum of all of these terms.</p>
<p>These sums of squares are summed within each frequency group, and dividing by TSS is the proportion of variance explained by that frequency group.</p>
<div class="section" id="technical-notes">
<h2>Technical Notes<a class="headerlink" href="#technical-notes" title="Permalink to this headline">¶</a></h2>
<p>The above regression is equivalent to the following in R:</p>
<div class="code R highlight-default"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">&lt;-</span> <span class="n">cbind</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">G</span><span class="p">)</span>
<span class="n">FORMULA</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;~&quot;</span><span class="p">,</span> <span class="n">paste</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">&quot; + &quot;</span><span class="p">))</span>
<span class="n">xx</span> <span class="o">=</span> <span class="n">summary</span><span class="p">(</span><span class="n">aov</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">formula</span><span class="p">(</span><span class="n">FORMULA</span><span class="p">),</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)))</span>
</pre></div>
</div>
<p>The regression method was implemented using the <a class="reference external" href="https://www.gnu.org/software/gsl">GNU Scientific Library</a></p>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Eyre-Walker, A. (2010). Evolution in health and medicine Sackler colloquium: Genetic architecture of a complex trait and its implications for fitness and genome-wide association studies. Proceedings of the National Academy of Sciences, 107 Suppl 1(suppl 1), 1752–1756. <a class="reference external" href="http://doi.org/10.1073/pnas.0906182107">http://doi.org/10.1073/pnas.0906182107</a></li>
<li>Lohmueller, K. E. (2014). The impact of population demography and selection on the genetic architecture of complex traits. PLoS Genetics, 10(5), e1004379. <a class="reference external" href="http://doi.org/10.1371/journal.pgen.1004379">http://doi.org/10.1371/journal.pgen.1004379</a></li>
<li>Sanjak, J., Long, A. D., &amp; Thornton, K. R. (2016). The Genetic Architecture of a Complex Trait is more Sensitive to Genetic Model than Population Growth.</li>
<li>Simons, Y. B., Turchin, M. C., Pritchard, J. K., &amp; Sella, G. (2014). The deleterious mutation load is insensitive to recent population history. Nature Genetics, 46(3), 220–224. <a class="reference external" href="http://doi.org/10.1038/ng.2896">http://doi.org/10.1038/ng.2896</a></li>
<li>Uricchio, L. H., Witte, J. S., &amp; Hernandez, R. D. (2015). Selection and explosive growth may hamper the performance of rare variant association tests. bioRxiv, 015917. <a class="reference external" href="http://doi.org/10.1101/015917">http://doi.org/10.1101/015917</a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Estimating cumulative <img class="math" src="../_images/math/aeb4b64115fe29aa42095dc692f5b5fb43dac071.png" alt="V(A)"/> in simulations of quantitative traits.</a><ul>
<li><a class="reference internal" href="#technical-notes">Technical Notes</a></li>
<li><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="../examples/copypops.html" title="previous chapter">Copying populations in memory</a></li>
      <li>Next: <a href="plugins.html" title="next chapter">Overview</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pages/vareg.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, KRT.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
      |
      <a href="../_sources/pages/vareg.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>